### ISSUE TITLE
Feature request

### ISSUE THREAD

**Issue by Monsterovich** [wormt/bcc#4]
This can make bcc more useful.
- c #defines

```
#define tan(x) FixedDiv(sin(x), cos(x))
```
- Short (structure) declarations

``` c++
struct Boss
{
int bosshp;
int bossarmor;
};

Boss a; // returns an error

struct Boss a; // right code
```
- Named enums

``` c++
enum BossType
{
    BOSSTYPE_REALMAN,
    BOSSTYPE_MONSTER
};

Bosstype type = BOSSTYPE_REALMAN;
```
- Ability to return structures & enums in functions

``` c++
Boss GetWeakBoss()
{
    Boss a;
    a.bosshp = 100;
    a.bossarmor = 200;
    return a;
}

BossType GetWeakBossType()
{
        return BOSSTYPE_MONSTER;
}
```

**Comment by positively-charged** [comment]
In the upcoming version of bcc, I added a C-like preprocessor. To be compatible with ACS, the `#define` and `#include` directives only work inside an `#if/#ifdef/#ifndef` block. The preprocessor is also case-sensitive, so keep that in mind.

Sample code:

```
#if 1

#define CONCAT( a, ... ) STRINGIZE( a ## __VA_ARGS__ )
#define STRINGIZE( a ) #a

script 1 open {
   Print( s: CONCAT( Hello, World! ) );
}

#endif
```

---

I added `typedef` to create type aliases, so you can make your declarations a little shorter. Right now, type aliases must end with "_T". It looks kind of ugly, but it simplifies the parsing and also forces a single naming convention onto everyone.

Sample code:

```
typedef struct Boss {
   int bosshp;
   int bossarmor;
} Boss_T;

Boss_T a;
struct Boss b;
```

---

Named enums are in the upcoming version. You still need to use the `enum` prefix, or use a type alias: 

```
typedef enum {
   BOSSTYPE_REALMAN,
   BOSSTYPE_MONSTER
} BossType_T;

BossType_T type = BOSSTYPE_REALMAN;
```

---

Named enums can be used as a return type:

```
function BossType_T GetWeakBossType() {
   return BOSSTYPE_MONSTER;
}
```

Structures must be returned by reference. At this time, array and structure references have limitations. Sample code:

```
function Boss_T& GetWeakBoss() {
   static Boss_T a = { 100, 200 };
   return a;
}
```

**Comment by positively-charged** [comment]
I introduced implicit type aliases for structs and enums, so the typedef is no longer necessary. If the name of the struct or enum is a type name, a type aliases referring to the struct or enum is implicitly created. I also made the type names look prettier by not requiring the underscore, although a lowercase letter is required instead:

```
struct BossT {
   int hp;
   int armor;
   enum BossTypeT {
      BOSSTYPE_REALMAN,
      BOSSTYPE_MONSTER
   } type;
};

BossT& GetWeakBoss() {
   static BossT a = { 100, 200, BOSSTYPE_MONSTER };
   return a;
}

BossTypeT GetWeakBossType() {
   return BOSSTYPE_MONSTER;
}
```

**Comment by Monsterovich** [comment]
So will it break compatibility with previous bcc versions?

**Comment by positively-charged** [comment]
Possibly. I decided to be as compatible with ACS as I can with the upcoming release. Some changes I made:
- Regions are replaced with namespaces.
- Format blocks are replaced with message-building functions, which are pretty much the same as format blocks but with a different syntax.
- Block scoping is only enabled by default in a namespace block.

These breaking changes are a one-time thing, but I think it's worth it. Overall, I'm really happy with the upcoming release.
